config/common/common.go

```diff
@@ -93,3 +103,102 @@ func GetFullyQualifiedIDFn(serviceProvider string, keyPairs ...string) tjconfig.
 		return path, nil
 	}
 }
+
+var referenceRules = [][]string{
+	{"resource_group$", "/azure/v1alpha1.ResourceGroup"},
+	{"subnet$", "/network/v1alpha1.Subnet"},
+}
+
+// AddCommonReferences adds some common reference fields.
+// This is a part of resource generation pipeline.
+func AddCommonReferences(references tjconfig.References, resource *schema.Resource, shortGroup string, nestedFieldNames []string) error {
+	for fieldName, s := range resource.Schema {
+		if s.Elem != nil {
+			e, ok := s.Elem.(*schema.Resource)
+			if ok {
+				nestedFieldNames = append(nestedFieldNames, fieldName)
+				if err := AddCommonReferences(references, e, shortGroup, nestedFieldNames); err != nil {
+					return err
+				}
+				if len(nestedFieldNames) > 0 {
+					nestedFieldNames = nestedFieldNames[:len(nestedFieldNames)-1]
+				}
+				continue
+			}
+		}
+
+		referenceName := prepareReferenceName(nestedFieldNames, fieldName)
+		referenceNameWithoutType, referenceKind := splitReferenceKindFromReferenceName(referenceName)
+		if referenceKind != "" {
+			referenceType := searchReference(referenceNameWithoutType)
+			if referenceType != "" {
+				if references == nil {
+					references = make(map[string]tjconfig.Reference)
+				}
+				if _, ok := references[referenceName]; !ok {
+					referenceType = prepareReferenceType(shortGroup, referenceType)
+					addReference(references, referenceKind, referenceName, referenceType)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func prepareReferenceName(elemNames []string, fieldName string) (result string) {
+	if len(elemNames) == 0 {
+		return fieldName
+	}
+	for _, e := range elemNames {
+		if result == "" {
+			result += e
+		} else {
+			result = fmt.Sprintf("%s.%s", result, e)
+		}
+	}
+	result = fmt.Sprintf("%s.%s", result, fieldName)
+	return
+}
+
+func splitReferenceKindFromReferenceName(fieldName string) (string, string) {
+	lastInd := strings.LastIndex(fieldName, "_")
+	if lastInd == -1 {
+		return "", ""
+	}
+	if fieldName[lastInd+1:] == NameReferenceKind || fieldName[lastInd+1:] == IDReferenceKind {
+		return fieldName[:lastInd], fieldName[lastInd+1:]
+	}
+	return "", ""
+}
+
+func searchReference(fieldName string) string {
+	for _, rule := range referenceRules {
+		r, _ := regexp.Compile(rule[0])
+		if r.MatchString(fieldName) {
+			return rule[1]
+		}
+	}
+	return ""
+}
+
+func prepareReferenceType(shortGroup, referenceType string) string {
+	if shortGroup == strings.Split(referenceType, "/")[1] {
+		referenceType = strings.Split(strings.Split(referenceType, "/")[2], ".")[1]
```

Comment:
We may cache the result of `strings.Split(referenceType, "/")`.

Reply:
Cached!